packages/vite/src/node/config.ts

```diff
@@ -205,18 +215,40 @@ export interface DevEnvironmentOptions {
   // fs: { strict?: boolean, allow, deny }
 }
 
-export type ResolvedDevEnvironmentOptions = Required<
-  Omit<DevEnvironmentOptions, 'createEnvironment'>
-> & {
-  // TODO: Should we set the default at config time? For now, it is defined on server init
-  createEnvironment:
-    | ((
-        name: string,
-        config: ResolvedConfig,
-      ) => Promise<DevEnvironment> | DevEnvironment)
-    | undefined
+function createDefaultClientDevEnvironment(
+  name: string,
+  config: ResolvedConfig,
+  context: CreateDevEnvironmentContext,
+) {
+  return new DevEnvironment(name, config, {
+    hot: context.ws,
+    watcher: context.watcher,
+  })
+}
+
+function createDefaultSsrDevEnvironment(
+  name: string,
+  config: ResolvedConfig,
+  context: CreateDevEnvironmentContext,
+): DevEnvironment {
+  return createNodeDevEnvironment(name, config, {
+    hot: createServerHotChannel(),
+    watcher: context.watcher,
+  })
 }
 
+function createDefaultDevEnvironment(
+  name: string,
+  config: ResolvedConfig,
+  context: CreateDevEnvironmentContext,
+): DevEnvironment {
+  return config.environments[name].consumer === 'client'
+    ? createDefaultClientDevEnvironment(name, config, context)
+    : createDefaultSsrDevEnvironment(name, config, context)
+}
```

Comment:
Before this PR, we were creating a `client`-like environment for any custom named environment that didn't define a `createEnvironment` function. I think this is a better default given that custom named environments are SSR-like by default.

@sheremet-va how do you see the default client environment if we deprecate `ws`? We can't have a `createClientHotChannel()` as we will need the HttpServer instance, no? If we want to avoid passing the whole Vite server, maybe we'll need to pass more things in the `CreateDevEnvironmentContext`.

I think moving this to the config face forces us to think what is needed to create an environment and play with the same rules for the default and custom ones.

Reply:
I'm sorry, I don't understand the question. We are deprecating the `ws` on the server, we are not removing the websocket server.

Reply:
Isn't it what I proposed? Pass down `ws`? https://github.com/vitejs/vite/pull/17791#discussion_r1696411143

Reply:
Yes, this is what we are doing now. But you also said "But make it optional so we don't need to deprecate it when we add ws: false)". IIUC, it is ok to have it always there, it won't need to be made optional as we still need to pass the `ws` so the client environment can use it.